pkg/kubelet/containerdshim/containerd_container.go

```diff
@@ -34,41 +37,49 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/dockershim"
 )
 
-const (
-	containerdCRIRoot = "/tmp/containerd-cri"
-	shimbindSocket    = "shim.sock"
-)
+// containerStore is used to store container metadata.
+// TODO: Consider to checkpoint ourselves or use containerd metadata store.
+var containerStore map[string]*runtimeapi.ContainerStatus = map[string]*runtimeapi.ContainerStatus{}
+var containerStoreLock sync.RWMutex
 
 // P0
 func (cs *containerdService) ListContainers(filter *runtimeapi.ContainerFilter) ([]*runtimeapi.Container, error) {
-	resp, err := cs.cdClient.List(gocontext.Background(), &execution.ListRequest{})
+	containerStoreLock.RLock()
+	defer containerStoreLock.RUnlock()
+	resp, err := cs.containerService.List(gocontext.Background(), &execution.ListRequest{})
 	if err != nil {
-		return nil, err
-	}
-	var criContainers []*runtimeapi.Container
-	for _, c := range resp.Containers {
-		criC, err := toCRIContainer(c)
-		if err != nil {
-			glog.Errorf("Failed to parse container %+v: %v", c, err)
-			continue
+		return nil, fmt.Errorf("failed to list containers from containerd: %v", err)
+	}
+
+	var containers []*runtimeapi.Container
+	for _, status := range containerStore {
+		container := statusToContainer(status)
+		// Set default state as exited, because any container without running
+		// containerd container is in exited state.
```

Comment:
The logic is that, if the shim thinks that there should be a container, but containerd told us that the container is not there, it means the container is stopped. :)